sympy/physics/quantum/qapply.py

```diff
@@ -1,206 +1,757 @@
+    --------------------------------------------
+
+    1.) The code relies on the auto-simplifications that Mul and all the composite objects constructors like
+    InnerProduct, OuterProduct, TensorProduct, Pow etc do: Extraction of commutative factors, contracting
+    of same factors to Powers etc, so these constructors may return their type, Mul or even other types.
+
+    2.) Mul(a,b) != a * b:
+    Mul(a, b, c) processes the arguments a, b, c (e.g. factoring out scalars, grouping same factors to powers)
+    but does not try to actually multiply the arguments, i.e. is not trying (a*(b*c)), (a*b) etc.
+    On the other hand a*b is executed using *-operator __mul__ for type(a), type(b):
+    E.g. for Kets and Bras, __mul__ is overloaded to form InnerProducts resp. OuterProducts types, and
+    e.g. IdentityOperator()*Ket(1)-> Ket(1)
+    (Examples: Mul(Bra(1), Ket(1)) returns unchanged. Bra(1)*Mul(Ket(1), Bra(1)) returns
+    Mul(Bra(1), Ket(1), Bra(1)), while Bra(1)*Ket(1)*Bra(1) returns Mul(InnerProduct(Bra(1),Ket(1)), Bra(1)).
+    So while the return type of Mul(a,b) will be Mul (or Zero) the return type of a*b completely depends
+    on the input types.
+
+    3.) Performance of Mul: refer to the remarks in code of Mul.flatten:
+    - avoid a*b*c if you don't need the multiplication and use Mul(a,b,c) instead
+    - instead of Mul(a,b)*Mul(c) use Mul(a,b,c)
+    - results of Mul(a,b,c) are cached
+
+
+    Remarks on qapply vs. Mul and "atomic" factors:
+    ----------------------------------------------
+
+    I will use the verbs 'to apply' and 'to multiply' equivalently as in this context they
+    mean the same and are both denoted by *.
+
+    So the purpose of qapply is to multiply factors that Mul doesn't know how to multiply, and
+    leave those that Mul knows to Mul. This necessarily includes the distribution of factors
+    over Add summands as far as required, which forces qapply to be a kind of workalike of expand().
+
+    For this purpose we have to classify all types of factors in two ways:
+
+    1. Does Mul know how to multiply this type with all other types, or does it not?
+
+       For efficiency, qapply would try to only handle those types that Mul doesn't
+       know how to multiply, and will leave all others to Mul. This also helps to avoid
+       contradicting results of Mul and qapply.
+       Of course passing factors to Mul makes only sense if there are more than one
+       to pass. So commutative factors are best. So it fits just nicely that in the
+       quantum context the only types Mul knows how to multiply are the scalars, i.e.
+       the complex numbers that may be pulled out.
+
+       So we would best use 'is_scalar' as predicate to send factors to Mul. Unfor-
+       tunately that doesn't exist, and the closest approximation 'is_complex' can
+       not be used: on most symbols, is_complex is undefined. The only predicate that
+       SymPy guarantees is 'is_commutative'!

```

Comment:
Perhaps what you are looking for here is `kind`:
```python
In [5]: (1 + sqrt(2)).kind
Out[5]: NumberKind

In [6]: Symbol('x').kind
Out[6]: NumberKind

In [7]: Matrix([[1]]).kind
Out[7]: MatrixKind(NumberKind)

In [9]: (y < 1).kind
Out[9]: BooleanKind
```
The kind system is incomplete but should probably be applied to quantum expressions to distinguish them from numbers:
```python
In [12]: Ket().kind
Out[12]: UndefinedKind
```
Ideally the kind here would be something like QuantumStateKind.

Reply:
I see. I will stick to 'commutative' to do the separations for the reasons explained in the comment. Actually I feel that QuantumStateKind  (or QuantumKind) wouldn't be a good idea in the long run: It would cement the idea that quantum objects are a different kind of animal than linear algebra objects - which is simply not the case and causes all types of trouble already by re-inventing the wheel for vectors / kets, matrices / operators, tensor products and TensorProducts in SymPy.

Reply:
> It would cement the idea that quantum objects are a different kind of animal than linear algebra objects

They are different in the context of SymPy. The kind system needs to differentiate all objects. The fact that multiplying a bra and a ket should do something special is something that Mul can only handle by understanding their kinds.

Note that I don't think it should be a generic QuantumKind. There is clearly a need to distinguish between operators and states. One way to think of it is whether it ever makes sense to add to objects of the same kind. It never makes sense to add a Bra and a Ket so they are not the same kind. Likewise though it doesn't make sense to add a vector and a ket at least not with SymPy's very spatial concept of vectors. Same goes for tensors etc.

The machinery used for handling the different kinds can potentially be shared (e.g. `Mul._kind_dispatcher`) but there definitely is a need to distinguish quantum operators and states from other "kinds" of object regardless of any parallels that may exist.